server/rating_service/rating_service.py

```diff
@@ -0,0 +1,382 @@
+                won_games = row[table.c.numGames] // 2
+            else:
+                won_games = row[table.c.winGames]
+
+            await conn.execute(
+                "INSERT INTO leaderboard_rating "
+                "(login_id, mean, deviation, total_games, won_games, leaderboard_id) "
+                "VALUES (%s, %s, %s, %s, %s, %s)",
+                player_id,
+                row[table.c.mean],
+                row[table.c.deviation],
+                row[table.c.numGames],
+                won_games,
+                self._rating_type_ids[rating_type.value],
+            )
+
+            return Rating(row[table.c.mean], row[table.c.deviation])
+
+    async def _create_default_rating(
+        self, conn, player_id: int, rating_type: RatingType
+    ):
+        default_mean = config.START_RATING_MEAN
+        default_deviation = config.START_RATING_DEV
+
+        rating_type_id = self._rating_type_ids.get(rating_type.value)
+        await conn.execute(
+            "INSERT INTO leaderboard_rating "
+            "(login_id, mean, deviation, total_games, won_games, leaderboard_id) "
+            "VALUES (%s, %s, %s, %s, %s, %s)",
+            player_id,
+            default_mean,
+            default_deviation,
+            0,
+            0,
+            rating_type_id,
+        )
+
+        return Rating(default_mean, default_deviation)
+
+    async def _persist_rating_changes(
+        self,
+        game_id: int,
+        rating_type: RatingType,
+        old_ratings: Dict[PlayerID, Rating],
+        new_ratings: Dict[PlayerID, Rating],
+        outcomes: Dict[PlayerID, GameOutcome],
+    ) -> None:
+        """
+        Persist computed ratings to the respective players' selected rating
+        """
+        self._logger.info("Saving rating change stats for game %i", game_id)
+
+        async with self._db.acquire() as conn:
+            for player_id, new_rating in new_ratings.items():
+                old_rating = old_ratings[player_id]
+                self._logger.debug(
+                    "New %s rating for player with id %s: %s -> %s",
+                    rating_type,
+                    player_id,
+                    old_rating,
+                    new_rating,
+                )
+                await conn.execute(
+                    "UPDATE game_player_stats "
+                    "SET after_mean = %s, after_deviation = %s, "
+                    "mean = %s, deviation = %s, scoreTime = NOW() "
+                    "WHERE gameId = %s AND playerId = %s",
+                    (
+                        new_rating.mu,
+                        new_rating.sigma,
+                        old_rating.mu,
+                        old_rating.sigma,
+                        game_id,
+                        player_id,
+                    ),
+                )
+
+                gps_rows = await conn.execute(
+                    select([gps_table.c.id]).where(
```

Comment:
I think the previous query will return this value, or there is a way to make it return the value. This is a pretty common thing you want so it's been optimized to be doable with only one query AFAIK

Reply:
Hm, using `returning` like in the link below (for postgres) tells my that "RETURNING is not supported by this dialect's statement compiler"...
https://docs.sqlalchemy.org/en/13/dialects/postgresql.html#insert-update-returning

Reply:
The best thing I though of so far is to find the row id first and then update the row in the second query, rather than the other way around. That way there is at least only one `WHERE` search that's not on the primary key.

Reply:
This is what I had in my memory but I guess it doesn't work for update statements. https://docs.sqlalchemy.org/en/13/core/connections.html#sqlalchemy.engine.ResultProxy.inserted_primary_key.

I think in this case we can update both table in a single expression:
https://www.mysqltutorial.org/mysql-update-join/

Reply:
Hm, but one is an update (to `game_player_stats`) and one is an insert (into `leaderboard_rating_journal`) which references the id of the updated row.

Reply:
Currently I am doing three queries and one search:
1. Select the `game_player_stats` row id based on the game data
2. Update that row (knowing it's id already)
3. Insert into `leaderboard_rating_journal`, referencing the id.

The following would have two queries and two searches:
1.Update the `game_player_stats` row (search for it based on game data)
2. Insert into  `leaderboard_rating_journal`, setting the id based on a `SELECT`.

Neither of them is as efficient as returning the updated id on the first query and I can't say whether the connection or the selection overhead is worse without experimenting on the real data.

Here's what the second query in the latter case would look like:
```python
journal_insert_sql = leaderboard_rating_journal.insert().values(
    leaderboard_id=rating_type_id,
    rating_mean_before=old_rating.mu,
    rating_deviation_before=old_rating.sigma,
    rating_mean_after=new_rating.mu,
    rating_deviation_after=new_rating.sigma,
    game_player_stats_id=select([game_player_stats.c.id]).where(
        and_(
            game_player_stats.c.playerId == player_id,
            game_player_stats.c.gameId == game_id,
        )
    )
)
```

Reply:
Yea that second one is what I was going to suggest next. Just making the `id` query a subquery of the insert statement.

Reply:
There is also the option of tracking and passing around game_player_stats ids (from game to rating service) but that's a pretty ugly solution just to save 1 query.

Reply:
I'll go for the subquery then.